### PR TITLE
Add support for raw-dylib with stdcall, fastcall functions

### PR DESCRIPTION
**Pull request by ricobbe**
Next stage of work for #58713: allow `extern "stdcall"` and `extern "fastcall"` with `#[link(kind = "raw-dylib")]`.

I've deliberately omitted support for vectorcall, as that doesn't currently work, and I wanted to get this out for review.  (I haven't really investigated the vectorcall failure much yet, but at first (very cursory) glance it appears that the problem is elsewhere.)

### CODE REVIEW

File: compiler/rustc_metadata/src/native_libs.rs [hunk 1]
```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
...
+                // Vectorcall is intentionally not supported at this time.
+                _ => {
+                    self.tcx.sess.span_fatal(
+                        item.span,
+                        r#"ABI not supported by `#[link(kind = "raw-dylib")]` on i686"#,
+                    );
+                }
+            }
+        } else {
+            DllCallingConvention::C
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/native_libs.rs:427]: This logic is going to normalize arbitrary irrelevant ABIs (e.g. `extern "msp430-interrupt"`) to `extern "C"`.

Also see some discussion on accepting ABIs on various targets in https://github.com/rust-lang/rust/pull/86231.

File: compiler/rustc_metadata/src/native_libs.rs [hunk 2]
```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
...
+    fn i686_arg_list_size(&self, item: &hir::ForeignItemRef<'_>) -> usize {
+        let argument_types: &List<Ty<'_>> = self.tcx.erase_late_bound_regions(
+            self.tcx
+                .type_of(item.id.def_id)
+                .fn_sig(self.tcx)
+                .inputs()
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        // stdcall on i686 never allocates less than 32 bits for an argument.
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/native_libs.rs:397]: Is this true for fastcall as well?

(Also I suspect that we already have the `i686_arg_list_size` logic somewhere in rustc, but can't check right now.)

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 3]
```diff
@@ -550,30 +550,23 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
+            dylib_table.entry(name).or_default().extend(lib.dll_imports.iter().cloned());
         }
     }
 
-    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
-    // what we should do if we have two DllImport values with the same name but different
-    // ordinals.
+    // Rustc already signals an error if we have two imports with the same name but different
+    // calling conventions (or function signatures), so we don't have pay attention to those
+    // when ordering.
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:559]: Is there a test for this case?

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 4]
```diff
@@ -550,30 +550,23 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
-                .map(|name| DllImport { name: *name, ordinal: None })
-                .collect::<Vec<_>>();
-            names.sort_unstable_by(|a: &DllImport, b: &DllImport| {
-                match a.name.as_str().cmp(&b.name.as_str()) {
-                    Ordering::Equal => a.ordinal.cmp(&b.ordinal),
-                    x => x,
-                }
+        .map(|(lib_name, import_set)| {
+            let mut imports = Vec::from_iter(import_set.into_iter());
+            imports.sort_unstable_by(|a: &DllImport, b: &DllImport| {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:566]:
```suggestion
            imports.sort_unstable_by_key(|a: &DllImport, b: &DllImport| {
```
is usually more convenient in cases like this.

File: compiler/rustc_codegen_llvm/src/back/archive.rs [hunk 5]
```diff
@@ -391,6 +391,21 @@ impl<'a> LlvmArchiveBuilder<'a> {
             ret
         }
     }
+
+    fn i686_decorated_name(import: &DllImport) -> CString {
+        let name = import.name.to_string();
```
Comment (petrochenkov) [inline compiler/rustc_codegen_llvm/src/back/archive.rs:396]:
```suggestion
        let name = import.name;
```
Symbols can be `format!`ed without allocating a new string.

File: compiler/rustc_codegen_llvm/src/back/archive.rs [hunk 6]
```diff
@@ -209,7 +209,7 @@ impl<'a> ArchiveBuilder<'a> for LlvmArchiveBuilder<'a> {
         let import_name_vector: Vec<CString> = dll_imports
             .iter()
             .map(if self.config.sess.target.arch == "x86" {
-                |import: &DllImport| CString::new(format!("_{}", import.name.to_string())).unwrap()
+                LlvmArchiveBuilder::i686_decorated_name
             } else {
                 |import: &DllImport| CString::new(import.name.to_string()).unwrap()
```
Comment (petrochenkov) [inline compiler/rustc_codegen_llvm/src/back/archive.rs:214]: Nit: it's a bit weird to return two closures from if/else, it's probably better to use one closure and process the `arch == "x86"` condition inside it.

File: src/test/run-make/raw-dylib-stdcall/lib.rs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+#![feature(raw_dylib)]
+
+#[repr(C)]
+struct S {
+    x: u8,
+    y: i32,
+}
+
+#[link(name = "extern", kind = "raw-dylib")]
+extern "stdcall" {
```
Comment (petrochenkov) [inline src/test/run-make/raw-dylib-stdcall/lib.rs:10]: Other newly supported ABIs also need test cases.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 2 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs:397]: It appears to be true: running run-make/raw-dylib-stdcall (replacing `stdcall` with `fastcall` throughout) does work.  As per comment below, I'll add tests for this case, and I'll update the comment.

And I wouldn't be at all surprised to learn that `i686_arg_list_size` is duplicate logic, but after digging around myself and consulting with various people, I was unable to find existing code that does the same thing.  If we can find the existing function, I'll be more than happy to use that instead.

File: src/test/run-make/raw-dylib-stdcall/lib.rs (hunk 7 above)
Comment (ricobbe) [reply src/test/run-make/raw-dylib-stdcall/lib.rs:10]: Added.

One question: I haven't changed the makefile, but there seems to be something wrong with my header command.  Could you have a look and tell me what I'm missing?  When I run the test with
```
x.py test -j 4 --build=i686-pc-windows-msvc src\test\run-make\raw-dylib-alt-calling-convention
```
it ignores the test, and I'm not sure why.  Deleting line 3 of the makefile allows it to run with the same command, so I've clearly gotten something wrong somewhere.

Comment (ricobbe) [reply src/test/run-make/raw-dylib-stdcall/lib.rs]: Fixed; apparently it needs to be `# only-i686-pc-windows-msvc` -- the "pc" is crucial, for some reason.

File: compiler/rustc_codegen_llvm/src/back/archive.rs (hunk 6 above)
Comment (petrochenkov) [reply compiler/rustc_codegen_llvm/src/back/archive.rs]: I meant `let import_name_vector: Vec<CString> = dll_imports.iter().map(new_single_function_for_both_64_bit_and_32_bit targets).collect();`

File: src/test/run-make/raw-dylib-alt-calling-convention/Makefile [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# Test the behavior of #[link(.., kind = "raw-dylib")] with alternative calling conventions.
+
+# only-i686-pc-windows-msvc
```
Comment (petrochenkov) [inline src/test/run-make/raw-dylib-alt-calling-convention/Makefile:3]: I suppose this should be testable on all MSVC targets since we are normalizing stdcall/fastcall/etc to C on everything except for `i686`?
(This kind of depends on decisions in https://github.com/rust-lang/rust/pull/86231 again.)

File: compiler/rustc_metadata/src/native_libs.rs (hunk 1 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs:427]: I read through the comments on the linked PR, and I'm a bit confused about the desired change here.  Should I signal an error on other cpu architectures for anything other than C, cdecl, stdcall, fastcall, and maybe vectorcall?  Or just signal an error for anything except C?  (MSVC ignores the other calling conventions on those platforms, but it's not clear that we should follow suit.)

It's probably worth pointing out that this function should only ever run on windows-msvc targets; I'll add a comment (or, better, an assertion) to that effect.

Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs:427]: Also, I notice that rustc supports an i586-pc-windows-msvc target.  Should we support stdcall & fastcall there also?

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 3 above)
Comment (ricobbe) [reply compiler/rustc_codegen_ssa/src/back/link.rs:559]: If both definitions are in the same scope, it results in E0428 ("the name `mumble` is defined multiple times").  I wasn't able to find a test for the case where the definitions have different calling conventions, but [test/ui/issues/issue-28472.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/issues/issue-28472.rs) comes close; it's two `extern "C"` functions with the same name and signature.

If the definitions are in different blocks (say, one inside a function and one at module top-level), then rustc just issues a lint warning: "`mumble` redeclared with a different signature;" this is covered by [test/ui/lint/clashing-extern-function.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/clashing-extern-fn.rs).  Do we need to upgrade this to an error?

File: src/test/run-make/raw-dylib-alt-calling-convention/Makefile (hunk 8 above)
Comment (ricobbe) [reply src/test/run-make/raw-dylib-alt-calling-convention/Makefile:3]: Well, the test passes when run in x86_64-pc-windows-msvc, though I'd like additional clarity on what we want to do with the various calling conventions on non-i686(/i586) platforms before relaxing the `only` header command.

File: compiler/rustc_codegen_llvm/src/back/archive.rs (hunk 6 above)
Comment (ricobbe) [reply compiler/rustc_codegen_llvm/src/back/archive.rs]: Changed; I was just lifting the architecture test out of the loop since it's invariant, but I can't imagine that perf matters that much here.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 2 above)
Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs]: The closest logic that I could find is `compiler\rustc_target\src\abi\call\x86.rs`, but it looks like the mangling logic doesn't care what is passed in registers or somewhere else, it just sums up the sizes.

Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs]: You are correct that the mangling logic doesn't differentiate between arguments passed on the stack and arguments passed in registers.  I haven't seen that documented anywhere, but the fact that this PR's run-make test passes provides pretty solid evidence for it.  😃 

I have a vague recollection that the stdcall convention requires the caller to allocate but not initialize stack space for arguments passed in registers, in order to give the called function a place to spill those registers if necessary, but I could be dreaming that up.  Either way, I think we've addressed this point -- if we find a better pre-existing function, then we can start another conversation and I'll address it then.

Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs]: Ugh, I looked through some code in LLVM and found at least two instances of this logic:
- `addByteCountSuffix` in `llvm-project\llvm\lib\IR\Mangler.cpp`  - aligns to 32 bit (or pointer size in general) instead of using it as a minimum like this PR, comment (incorrectly?) says that the result is a number of words rather than the number of bytes.
- `MangleContext::mangleName` in `llvm-project\clang\lib\AST\Mangle.cpp` - seems to do the same thing, but with some messy divisions and multiplications to get the number of bytes (again) instead of the number of words.

Looks like the alignment vs minimum mismatch is a thing to fix.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs:559]: I think we shouldn't touch the lint, but report a new hard error here if mismatching declarations are used with raw-dylib feature specifically.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 9]
```diff
@@ -550,30 +550,21 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
-                }
-            });
-            (lib_name, names)
+        .map(|(lib_name, import_set)| {
+            let mut imports = Vec::from_iter(import_set.into_iter());
+            imports.sort_unstable_by_key(|x: &DllImport| x.name.as_str());
+            (lib_name, imports)
         })
         .collect::<Vec<_>>();
     result.sort_unstable_by(|a: &(String, Vec<DllImport>), b: &(String, Vec<DllImport>)| {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:570]:
```suggestion
    result.sort_unstable_by_key(...
```

File: compiler/rustc_metadata/src/native_libs.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs:427]: >Also, I notice that rustc supports an i586-pc-windows-msvc target. Should we support stdcall & fastcall there also?

It should be automatically supported, `i586` is basically the same target as `i686` just with a different base CPU (controlling the set of available CPU features), see e.g. the target definition in `compiler\rustc_target\src\spec\i586_pc_windows_msvc.rs`.

Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs:427]: >I read through the comments on the linked PR, and I'm a bit confused about the desired change here.

I'm not entirely sure yet.
I suspect that MSVC supported the "legacy" conventions on x64 for compatibility reasons, to ease portability of huge amount of pre-existing code.
We don't have such a problem, so for us it would be better to reject all the "legacy" conventions on x64.

Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs:427]: Let's accept only `Abi::C` and `Abi::Win64` here for a start, and produce the "ABI not supported" error for anything else.

Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs:427]: (Note that vectorcall is a thing on x64, and it's different from the default C/win64, and I have no idea how it's mangled there, but this PR doesn't support vectorcall, so it's not a problem right now.)

File: src/test/run-make/raw-dylib-alt-calling-convention/Makefile (hunk 8 above)
Comment (petrochenkov) [reply src/test/run-make/raw-dylib-alt-calling-convention/Makefile:3]: Ok, let's leave this as is in accordance with https://github.com/rust-lang/rust/pull/86419#discussion_r660780299.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 1 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs:427]: Re vectorcall: interesting; I wasn't aware that was a thing on 64-bit.  Good to know!

For context, I decided not to support vectorcall on 32-bit because there appear to be deeper problems with it than just getting the name mangling correct.  Making the obvious modification here to allow vectorcall and running the PR's run-make test (with, effectively, `s/stdcall/vectorcall/g;`) didn't work, but in an interesting way: it all links and runs, but the output indicates that the called function isn't seeing the same parameter values that the calling function provides.  So my hunch is that there are other issues with vectorcall in rustc that are outside the scope of this PR.  My primary motivation with the raw-dylib work is facilitating interop with the Windows API, and effectively all of those functions are stdcall on x86-32, so it just didn't seem worth it to dig into the vectorcall issues further.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 9 above)
Comment (ricobbe) [reply compiler/rustc_codegen_ssa/src/back/link.rs:570]: Yeah, I've tried to get that to work, but I can't make the lifetimes work (unless I copy the strings, but that seems wasteful).

```
573 |     result.sort_unstable_by_key(|x: &(String, Vec<DllImport>)| &x.0);
    |                                     -                        - ^^^^ returning this value requires that `'1` must outlive `'2`
    |                                     |                        |
    |                                     |                        return type of closure is &'2 std::string::String
    |                                     let's call the lifetime of this reference `'1`
```

Lifting the closure to a named function with lifetime parameters doesn't work either:
```
    fn extract_key<'a, 'b: 'a>(x: &'b (String, Vec<DllImport>)) -> &'a str { &x.0 }

    result.sort_unstable_by_key(extract_key);
```
results in an error saying only "higher-ranked subtype error" at the call site.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 2 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs]: Yeah, I saw the comment in `addByteCountSuffix`, or one just like it.  I can only assume, based on experience, that the use of "words" there is incorrect -- using size in words instead of bytes in the mangled names results in an import library that doesn't link correctly against the DLL produced by MSVC.  The details of stdcall that I've been able to find at docs.microsoft.com are suspiciously silent on this point, so I kinda have to take the empirical results as definitive.  I'd personally prefer to have better clarity in the spec on this point, but 🤷 .

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 9 above)
Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs:570]: Ok.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 1 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs:427]: Also added `Abi::System` for x86_64, as that appears to be a synonym for `Abi::C` (at least on x86_64) that devs will expect to be able to use.

File: compiler/rustc_metadata/src/native_libs.rs (hunk 2 above)
Comment (ricobbe) [reply compiler/rustc_metadata/src/native_libs.rs]: > Looks like the alignment vs minimum mismatch is a thing to fix.

Well, after further experimentation, I agree that `max(4, layout.size.bytes_usize())` is not correct, and I've updated this, although I haven't pushed yet.

Based on what I've observed, it seems that stdcall expects that the space for a given argument is the smallest multiple of 4 bytes into which that argument can fit; this is true even for types like `f64` that might under other circumstances be aligned to the nearest 8-byte boundary.

The alignment information that I get from the `Layout` struct doesn't really help here.  As one example, the `Layout` value for `u8` indicates an alignment of 1 byte (both abi and preferred), which doesn't match stdcall: `fn f(x: u8, y: u8)` has an argument size of 8 bytes, not 2.

Unfortunately, messy code is probably unavoidable -- best I've got so far is `(layout.size.bytes_usize() + 3) & !3`, with a comment explaining what's going on.

Beyond that, is there anything else you want to see here?

Comment (petrochenkov) [reply compiler/rustc_metadata/src/native_libs.rs]: >Beyond that, is there anything else you want to see here?

No, the `min` was the only thing that looked incorrect here.

File: compiler/rustc_metadata/src/native_libs.rs [hunk 10]
```diff
@@ -396,4 +384,55 @@ impl Collector<'tcx> {
...
+    fn i686_arg_list_size(&self, item: &hir::ForeignItemRef<'_>) -> usize {
+        let argument_types: &List<Ty<'_>> = self.tcx.erase_late_bound_regions(
+            self.tcx
+                .type_of(item.id.def_id)
+                .fn_sig(self.tcx)
+                .inputs()
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        argument_types.iter().fold(0, |accum: usize, ty: Ty<'_>| {
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/native_libs.rs:397]: It would be simpler to use `Iterator::sum` https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.sum instead of `fold` here.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 11]
```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
-    // ordinals.
+    // Check for multiple imports with the same name but different calling conventions or
+    // (when relevant) argument list sizes.  Rustc only signals an error for this if the
+    // declarations are at the same scope level; if one shadows the other, we only get a lint
+    // warning.
+    for (library, imports) in &dylib_table {
+        let mut import_table: FxHashMap<Symbol, &DllImport> = FxHashMap::default();
+        for import in imports {
+            if let Some(old_import) = import_table.insert(import.name, &import) {
+                if import != old_import {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:568]: `DllImport` contains a `Span` now, so you cannot use derived `PartialEq`/`Hash` anymore, spans are almost always going to be different.
`import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 12]
```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
+                    sess.span_fatal(
+                        import.span,
+                        &format!(
+                            "multiple inconsistent definitions of external function `{}` from library `{}`",
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:572]: It's better to explicitly say that the functions have different ABIs, otherwise it's not clear how exactly they are inconsistent.

**Comment by petrochenkov** [comment]
@ricobbe 
Could you squash commits after addressing the remaining comments?

**Comment by ricobbe** [comment]
> Could you squash commits after addressing the remaining comments?

Will do, sure.  For my own understanding, what is preferred practice on this?  Squash at the end before completing the PR, or squash as I go?

**Comment by petrochenkov** [comment]
@ricobbe 
>For my own understanding, what is preferred practice on this? Squash at the end before completing the PR, or squash as I go?

Both ways work for me.
The important part is to have a reasonable git history immediately before merging.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 11 above)
Comment (ricobbe) [reply compiler/rustc_codegen_ssa/src/back/link.rs:568]: Ok, it looks like I'm sort of stuck here.  As far as I can determine, something in the query system requires `NativeLib` to implement `HashStable`.  (I'm sort of guessing here because the error happens somewhere in a big stack of macro invocations, and I'm having trouble figuring out where in this stack this requirement is imposed.)  Therefore, adding `Span` information to `NativeLib` (even indirectly, as I have) doesn't seem to be an option.

The `Span` is only necessary so I could give some sort of source location information in the error message that we signal if we encounter multiple declarations of the same external function with inconsistent calling conventions.  Do we want to settle for issuing a diagnostic without location info?  (I'm not wild about this choice.)  Alternatively, is there a better place in the compiler to detect this case and issue the error?  I would expect that a big difficulty with doing this before we've created the `NativeLib` structs is that we don't have easy access to information about all of the external declarations in one place.

> `import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

I think this gets into a question about how much future proofing we should be doing here.  The RFC does call for eventual support of linking by ordinal instead of by name, and `DllImport` contains the `ordinal` field to support this in the future.  At the moment, nobody ever reads that field.  Going forward, I have no idea if/when I'm going to get around to implementing that specific bit of functionality; our primary concern is to make interoperability with the Windows API easier, and support for ordinals isn't very important for that.  I'm starting to wonder if I should just drop the field (but otherwise leave `DllImport` intact) rather than including only half-hearted support for ordinals.

Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs:568]: >something in the query system requires `NativeLib` to implement `HashStable`

Derived `Hash` implementation (including `Span`) works ok for the purposes of query/incremental systems, it'll just add some dependency on spans, so it can be kept.
The important thing is to not use for logic like duplicate detection here.

>I think this gets into a question about how much future proofing we should be doing here. The RFC does call for eventual support of linking by ordinal instead of by name

Then `FxHashMap<Symbol, DllCallingConvention>` will turn into `FxHashMap<Symbol, (DllCallingConvention, Option<u16>)>`.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 13]
```diff
@@ -464,30 +467,44 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
...
+        }
+    }
+
+    // Check for multiple imports with the same name but different calling conventions or
+    // (when relevant) argument list sizes.  Rustc only signals an error for this if the
+    // declarations are at the same scope level; if one shadows the other, we only get a lint
+    // warning.
+    for (library, imports) in &dylib_table {
+        let mut import_table: FxHashMap<Symbol, DllCallingConvention> = FxHashMap::default();
+        for import in imports {
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:480]: "multiple definitions" are reported by walking an `FxHashSet`, so the walking order is unpredictable and `.stderr` files in the test suite can be different causing the CI failure.

Perhaps this logic needs to be placed after the sorting code below.

Comment (ricobbe) [reply compiler/rustc_codegen_ssa/src/back/link.rs:480]: Yup, just figured this out.  Working on a fix now; should only take an hour or so, barring interrupts from other stuff.

**Comment by joshtriplett** [comment]
This appears to have added a failing test:
```
8    = note: see issue #58713 <https://github.com/rust-lang/rust/issues/58713> for more information
9 
9 
10 error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
-   --> $DIR/multiple-definitions.rs:8:5
12    |
12    |
- LL |     fn f(x: i32);
-    |     ^^^^^^^^^^^^^
+ LL |         fn f(x: i32);
15 
16 error: aborting due to previous error; 1 warning emitted
17
```

**Comment by RalfJung** [comment]
> This appears to have added a failing test:

How is that even possible...?

**Comment by fee1-dead** [comment]
This is weird, both #86922 and #86857 encountered this but they seem unrelated.

#87075 *fails* again.

**Comment by RalfJung** [comment]
Looks like we have different output on different targets?
But bors wouldn't have landed unless the stderr file is correct for *all* targets...

**Comment by RalfJung** [comment]
Is it possible that something depends on hash table ordering here (like the code that determines which of the "multiple definitions" the span is pointing at)?

**Comment by ricobbe** [comment]
I'm finding it a bit hard to follow the flurry of communications about this PR that happened over the weekend.  What's the current status of this work--is the multiple-definitions test still failing in automation?

> Is it possible that something depends on hash table ordering here (like the code that determines which of the "multiple definitions" the span is pointing at)?

An earlier version of this PR did indeed have this problem.  I checked in a fix before the PR was merged, but it's possible that there's still some nondeterminism left.
